flows/photometry.py

```diff
@@ -232,22 +238,93 @@ def photometry(fileid, output_folder=None, attempt_imagematch=True):
 		edge[(0,-1),:] = True
 		edge[:,(0,-1)] = True
 		curr_star -= nanmedian(curr_star[edge])
-		curr_star /= np.max(curr_star)
+		curr_star /= nanmax(curr_star)
 
 		ypos, xpos = np.mgrid[:curr_star.shape[0], :curr_star.shape[1]]
 		gfit = gfitter(g2d, x=xpos, y=ypos, z=curr_star)
 
 		fwhms[i] = gfit.x_fwhm
+		gauss_pos[i, :] = [xmin + gfit.x_mean, ymin + gfit.y_mean]
+
+		# Calculate rsq
+		sstot = ((curr_star - curr_star.mean()) ** 2).sum()
+		sserr = (gfitter.fit_info['fvec'] ** 2).sum()
+		rsqs[i] = 1.0 - sserr/sstot
 
 	masked_fwhms = np.ma.masked_array(fwhms, ~np.isfinite(fwhms))
-	fwhm = np.mean(sigma_clip(masked_fwhms, maxiters=20, sigma=2.0))
+	masked_rsqs = np.ma.masked_array(rsqs, ~np.isfinite(rsqs))
+
+	# Create plot of target and reference star positions from 2D Gaussian fits.
+	fig, ax = plt.subplots(1, 1, figsize=(20, 18))
+	plot_image(image.subclean, ax=ax, scale='log', cbar='right', title=target_name)
+	ax.scatter(gauss_pos[:, 0], gauss_pos[:, 1], c='r', marker='o', alpha=0.6, label='2D Gauss')
+	ax.scatter(references['pixel_column'], references['pixel_row'], c='g', marker='o', alpha=0.6, label='References')
+	ax.scatter(target_pixel_pos[0], target_pixel_pos[1], marker='+', s=20, c='r', label='Target')
+	ax.legend()
+	fig.savefig(os.path.join(output_folder, 'positions_2dgauss.png'), bbox_inches='tight')
+	plt.close(fig)
+
+	# Use R^2 to more robustly determine initial FWHM guess.
+	# This cleaning is good when we have FEW references.
+	min_fwhm_references = 2
+	min_references = 6
+	min_references_now = min_references
+	rsq_min = 0.15
+	rsqvals = np.arange(0.95, rsq_min, -0.15)
+	fwhm_found = False
+	min_references_achieved = False
+
+	# Clean based on R^2 Value
+	while not min_references_achieved:
+		for rsqval in rsqvals:
+			mask = (masked_rsqs >= rsqval) & (masked_rsqs < 1.0)
+			nreferences = np.sum(np.isfinite(masked_fwhms[mask]))
+			if nreferences >= min_fwhm_references:
+				_fwhms_cut_ = np.mean(sigma_clip(masked_fwhms[mask], maxiters=100, sigma=2.0))
+				logger.debug('R^2 >= %.2f: %d stars with mean FWHM = %.2f',
+					rsqval, nreferences, _fwhms_cut_)
+				if not fwhm_found:
+					fwhm = _fwhms_cut_
+					fwhm_found = True
+			if nreferences >= min_references_now:
+				references = references[mask]
+				min_references_achieved = True
+				break
+
+		if min_references_achieved:
+			break
+		min_references_now -= 2
+		if fwhm_found and min_references_now < 2:
+			break
+		elif not fwhm_found:
+			raise Exception("Could not estimate FWHM")
+		logger.debug('%d, %d, #refs = %d', min_references_now, min_fwhm_references, nreferences)
+
 	logger.info("FWHM: %f", fwhm)
 	if np.isnan(fwhm):
 		raise Exception("Could not estimate FWHM")
 
+	# if minimum references not found, then take what we can get with even a weaker cut.
+	# TODO: Is this right, or should we grab rsq_min (or even weaker?)
+	min_references_now = min_references - 2
```

Comment:
Regarding this code block, are you in fact not just picking the `min_references_now` largest rsquears's? I don't think there is a need for a loop, which potentially could go into an infinite loop?

Reply:
So the loop is there to try to get as many stars as possible, we soft reset the minimum references, and then try again with the lowest R^2 value. If it doesn't work, we decrease R^2 and minimum references until we do finally reach enough. This should exit gracefully since min_references and the R^2 minimum are lowered in each loop and will quickly go to zero. Then nreferences will be >= min_references and it will exit.

It actually cannot go infinite since min_references will keep decreasing to zero, and n references is finite.

Reply:
We can try to exit early by checking that there are no R^2 values between 0 and 1, and raising and exception in that case. But in reality this outcome is just a few iterations away and we don't really need to explicitly check for it.

Reply:
But since you are just lowering the limit until you get enough stars, it would be quicker to just calculate which limit would give you that required number of stars... Or just sorting the stars by rsquare and picking the number you require

Reply:
Something along the lines of this?
```
mask = (masked_rsqs >= rsq_min) & (masked_rsqs < 1.0)
mask = np.argsort(masked_rsqs[mask]) # Sort by largest rsquare
mask = mask[:min(min_references, len(mask))] # Only take the min_references largest
```

Reply:
**TL;DR**: Let's check the minimum rsq_val that gives us at least 2 stars. If it's >=0.15 we are happy, if it's <0.15 but >0.00, we should print a warning and try, if it's <0.01, we should raise. Like you said, we don't need a loop for this.

------- 
Well we are doing it greedily. Meaning that we change what it means to get "enough stars" at the same time as we lower R^2. 

If we do as you suggest (which is more efficient), then we need to preselect the minimum acceptable R^2 value. In my opinion, that depends on the minimum acceptable stars said value gives. (Perhaps AIC/BIC would work better here anyway.) 

I set it up here so that we first try to get at least 2 stars at our previous minimum R^2 = 0.2. (But higher is better, this all happens in the previous loop)

Then, in this loop, we soft reset and try to get at least 4 stars from R^2 = 0.15. If we don't, then we lower to 3 and 0.08.  then 2 and 0.01 (worst case minimum acceptable scenario, but we only want this if 0.15 and 0.08 don't work.) 

 (Maybe it's easier to pre-calculate if you can code it that way?) But I think the pre-sort would work.

. Otherwise, we could just go to the R^2 value that will give us 2 stars. But we want to know whether there is an R^2 value that will give us more than 2 stars.

Reply:
I guess it is just unclear to me what the goal is here...

If it is to get as many reference stars as possible which have a rsquare over 0.15, why not just take all the ones which are above 0.15 and then warn/stop if there are not enough?

If it is to get a set number of reference stars, then just sort by rsquare and pick the desired number of stars from the top of that list, and then warn/stop if there are not enough?

Reply:
I agree it's hacky but the goal is to filter out low R^2 value stars, if possible, if not, go lower until we reach just enough stars. I think it's going to work better in the wcs checking version I mentioned. 

We basically want both high R^2 and high number of stars. We can't really set the limit independently. The more we lower R^2, the higher chance of bad stars. But if we just do a high R^2 cutoff, we risk missing stars that we could have used, and having too few for the zero point statistics.

Just to illustrate:
If we can have 20 stars, we don't really want to use R^2 < 0.5. To minimize risk of bad stars.
If we can have 5 stars, we don't really want to use R^2 < 0.2, to minimize risk of bad stars but maximize number of stars.
if we can only have 3 stars, we are willing to go even lower in R^2. and so on.

I think we should do as you suggest with pre-computing. But this was my hacky solution to what I want to accomplish above. I'm sure there's a better way!

Reply:
What I suggest is adding "improving the false star rejection" as an issue, and working on that when we get a chance. I think using some other goodness of fit criteria might work better for us. I'm not really willing to work more on this hacky solution at the moment as I think it will be redundant when we do so.